### PR TITLE
Accessibility enhancements: catching a11y errors with WAVE

### PR DESCRIPTION
**Pull request by felix91gr**
Addesses some of the issues shown at #171 (we're skipping the contrast errors for this PR). Checks were done using the [WAVE tool](https://wave.webaim.org/), in particular, with the [Firefox add-on](https://addons.mozilla.org/en-US/firefox/addon/wave-accessibility-tool/).

Further accessibility PRs will focus on interactivity (using the Firefox Accessibility Inspector) and contrast (using WAVE again).

**Error Checklist:**

- [x] Missing image alt text
- [x] Missing document language
- [x] Missing form labels
- [x] Empty buttons
- [x] 1x Empty table header
- [x] 1x Broken ARIA reference

**Sub-checklist for contrast errors:**

(_These will be addressed in a later PR_)

- Blue Buttons (Light theme) 
- Other Blue UI elements (Light theme) 
- Gray text over white background (Light theme) 
- Some tag colors have low contrast (Light theme) 
- Contrast Errors (Dark theme) (TODO: needs to be expanded upon)

**Warning Checklist:**

Warnings are sometimes relevant, depending on the use-case.

- [x] 2x Orphaned form labels
- [x] 1x Missing fieldset
- [x] Skipped heading levels
- [x] ~1x Possible table caption~ Deemed not relevant because WAVE is thinking that the "loading" row of the playlist table might be a caption (but isn't).
- [x] ~1x No page regions~ We're skipping this one because the page is so simple it doesn't need regions. Or maybe it does! Please answer to [this comment](https://github.com/azlux/botamusique/pull/180#issuecomment-653660451) if you think the page might need some regions :3

### CODE REVIEW

**Comment by felix91gr** [comment]
@TylerVigario I can't seem to be able to set the language in the final HTML. As you can see, I have this in the first line of the document:

```html
<!DOCTYPE html lang="en">
```

Yet when it gets rendered with node, this HTML tag gets changed to:

```html
<!doctype html>
```

So I'm guessing there are hooks at the Node level from which I'm supposed to set the document language. Is that correct?

**Comment by TylerVigario** [comment]
HTML5 changed the syntax for that.

Try something like this

```HTML
<!DOCTYPE html>
<html lang="en">
```

**Comment by felix91gr** [comment]
That worked! Thanks, @TylerVigario :3

**Comment by felix91gr** [comment]
The blue buttons in the light theme have contrast problems. 

Currently I've set darkened the button color, and checked that it passed the contrast requirements with [this tool](https://webaim.org/resources/contrastchecker/).

I still have to add a darker color for the hover behavior (because now it transitions from dark to light, whereas all the other buttons in the Light theme go from light to dark), but I'd like to have some feedback for now. @TylerVigario or @TerryGeng, who do I ask for opinions on color theme? :P (Maybe it can be switched to a different color and not just a darker one like I did :3)

**Comment by felix91gr** [comment]
Hmm, I think I will leave the contrast issues for last, or for another PR, because they are much harder to spot. There are almost none in the initial UI; most of them appear after opening the music library, adding tags and adding music to the playlist.

File: web/templates/index.html [hunk 1]
```diff
@@ -83,25 +91,29 @@ <h1>botamusique Web Interface</h1>
...
                     </tr>
                     <tr class="playlist-expand table-dark d-none">
                         <td colspan="4" class="text-center">
-                            <a class="text-muted" href="javascript:">See item <span class="playlist-expand-item-range"></span> on the playlist.</a>
+                            <a class="text-muted" href="javascript:">See item <span
+                                    class="playlist-expand-item-range"></span> on the playlist.</a>
                         </td>
                     </tr>
                     <tr class="playlist-item-template d-none">
                         <th scope="row" class="playlist-item-index d-none d-md-table-cell"></th>
```
Comment (felix91gr) [inline web/templates/index.html:111]: @TylerVigario @TerryGeng what does this table header do?

I've seen its content change from nothing, to 1, to 2, etc, when adding songs to the playlist, and down from 3 to 2 and from 2 to 1 when removing songs from the playlist. What is manipulating this header, and what is its purpose? Do you know?

I ask because if it is just like a global variable or something like that, I can mark it as `aria-hidden` for screen reader users.

Comment (TerryGeng) [reply web/templates/index.html:112]: If you have more than 10 items on the playlist (like sometimes I will add about ~100 songs to the playlist), the web interface will only display 10 items of them to avoid making the entire webpage too long.

I'm actually thinking about pagination like what it is like in the library section. I think I will implement it in the next few weeks after I manage to get #158 done.

Comment (felix91gr) [reply web/templates/index.html:112]: > the web interface will only display 10 items of them to avoid making the entire webpage too long.

Of course :3 that's good.

But what I mean to ask is what does this `<th>` tag do. Is it a counter of the items in the playlist? :)

> I'm actually thinking about pagination like what it is like in the library section

Ohh, actually now that you mention it, that would be sweet. I hadn't realized that they were implemented differently, in fact, I didn't even think to look at the library impl :D

Comment (TerryGeng) [reply web/templates/index.html:112]: I think this `<th>` is displayed as the index of items.
image

Comment (felix91gr) [reply web/templates/index.html:112]: I thought so too, but its `display` attribute is always `none`. I think it's being cloned every time one adds a song, and the clone's `display` attribute is set to `show`

Comment (felix91gr) [reply web/templates/index.html:112]: It's the one `<th>` that always has the class `playlist-item-index d-none d-md-table-cell`, even when there are no songs in the playlist :3

Comment (TerryGeng) [reply web/templates/index.html:112]: Oh. You mean this one _which is always hidden_. This is a template, every time a song is being added to the playlist, I clone this object and set its display attribute to `block`.

**Comment by felix91gr** [comment]
I have updated the description to exclude the color changes from this PR. Color changes are a special hurdle that deserves its own dedicated discussion (because the color palette is a big topic).

TLDR: will only address non-contrast errors + warnings in this PR :3

File: web/templates/index.html (hunk 1 above)
Comment (felix91gr) [reply web/templates/index.html:112]: Yaaaaaas!

Comment (felix91gr) [reply web/templates/index.html:112]: Okay, that means that everything's alright with regards to that header :)

Comment (felix91gr) [reply web/templates/index.html:112]: Context: [the tool I'm using](https://wave.webaim.org/) was complaining because it wants to see all table headers with text. This one header however, is alright by the accessibility standards, because its `display` is always `none`. This means that the tool is just confused.

This completes the first checklist :)

Comment (felix91gr) [reply web/templates/index.html:112]: PS: [here](https://webaim.org/techniques/tables/data) is the justification for wanting text in the table headers. It makes sense, it's just that WAVE (the tool) can't see that they are being hidden.

**Comment by felix91gr** [comment]
The PR is ready for review.

I still have two warnings to clear (_Possible table caption_ and _No page regions_), but the bulk of this PR is already done for if you want to take a look.

CC @TerryGeng @TylerVigario

**Comment by felix91gr** [comment]
I have added a dynamic ARIA label for the main Play/Pause button, because it changes behaviour depending on the state of the app :)

**Comment by felix91gr** [comment]
Okay, it is done. Last changes were:

1. :heavy_check_mark: I added JavaScript behaviour to address my last relevant `FIXME`. Basically the Play/Pause button changes its role depending on the state of the playlist and if there's music playing as well. I cannot reflect this in the HTML, so I had to go to JS to set the attribute manually whenever it changed roles.
2. :question: I checked the last two warnings that I had listed, and deemed them not important for our use-case:
    a.  **Possible table caption**. :heavy_check_mark: This is a warning that gets issued when you have some text that might be intended to be the caption for a table, but is not inside a `<caption>` tag. This is very useful to notice whenever you have a table and have captioned it visually, but not semantically. `<caption>` helps convey that meaning to blind users. In our case, the text the tool recognized as a possible caption, was just a row of the playlist table which is being used to convey that the playlist is loading. Since this is not a caption use-case, it is alright to ignore the warning.
    b. **No page regions** :question: We're skipping this one because the page is so simple it doesn't need regions. Or maybe it does? I will put the explanation for this warning below. @TerryGeng I need your feedback on this one.

---

### Warning: No Page Regions

Terry, here is the explanation for the No Page Regions warning. What do you think of it?

(Copied word by word from the WAVE tool:)

#### What It Means
No page regions or ARIA landmarks were found.

#### Why It Matters
Regions and ARIA landmarks identify significant page areas. Most web pages should have regions defined, particularly for the main content area.

#### How to Fix It
If the page has visual regions or significant page areas, ensure the regions are defined with header, nav, main, footer, etc. elements.

Relevant guidelines source: [2.4.1 Bypass Blocks (Level A)](https://webaim.org/standards/wcag/checklist#sc2.4.1)

**Comment by felix91gr** [comment]
I have deployed this on my server, I will be testing it now with a complete sound library and covering hopefully all the use cases :3 
So far it looks good, but there are some labels I have missed, and maybe a semantic change in hidden tags that could be handy. I will keep iterating over it, although I will leave the contrast errors for another PR :)

File: web/templates/index.html [hunk 2]
```diff
@@ -218,7 +218,7 @@ <h3 class="card-title">Filters</h3>
                             alt="A drawing of an empty box." />
                     </div>
                     <div id="library-item" style="display: none;" class="list-group-item library-item">
-                        <input hidden type="text" class="library-item-id" value="" aria-hidden="true" />
+                        <input hidden type="hidden" class="library-item-id" value="" />
```
Comment (felix91gr) [inline web/templates/index.html:221]: These changes are semantic. These are inputs that are being hidden using CSS, which is visually fine, but are actually still being shown to blind users. These should be of **type** `hidden` instead, since _semantically,_ they are not supossed to be seen.

**Comment by felix91gr** [comment]
Okay, I have solved the semantically hidden input issue. I want to keep testing this on my server, but so far it has looked fine.

I haven't seen any new errors with WAVE, so I think this might be good to go with regards to the accessibility errors.

There is one small problem that I want to solve that is not related to contrast (contrast will be addressed in a later PR), and that is the alt text for cover art. I think it's okay to show a default alt text for when there is no cover art, but I would like to show some text when there is. I'm thinking of just saying _"Cover art of <name_of_song>"_

Oh and the Page Regions question as well. That and the cover art alt text are the two remaining points I think.

File: web/templates/index.html [hunk 3]
```diff
@@ -107,7 +107,7 @@ <h1>botamusique Web Interface</h1>
                         </td>
                     </tr>
                     <tr class="playlist-item-template d-none">
-                        <th scope="row" class="playlist-item-index d-none d-md-table-cell"></th>
+                        <th scope="row" class="playlist-item-index d-none d-md-table-cell">1</th>
```
Comment (felix91gr) [inline web/templates/index.html:110]: I have added this because this header is empty only when the page has an empty playlist and is reloaded. After emptying the playlist, this is how the header looks. It is more consistent this way, and it also gets rid of a particular error.

Although we might want to hide this with an ARIA tag instead. A hidden header should not be shown to anyone, really. This one is just a template.

Comment (felix91gr) [reply web/templates/index.html:110]: If I understand correctly, @TerryGeng, everything inside the `<tr class="playlist-item-template d-none">` tag is just a template, right? If so, I can fix this issue easily: I can set the `<tr>` tag to be hidden from view for blind users.

Comment (felix91gr) [reply web/templates/index.html:110]: And that would not need the templating code to change, because I would change the attributes of the container, and not of any of the templates. If that makes sense.

Comment (felix91gr) [reply web/templates/index.html:110]: Okay I added an `aria-hidden="true"` attribute to the parent as well.

**Comment by felix91gr** [comment]
Last update for today: I think I need an answer to https://github.com/azlux/botamusique/pull/180#issuecomment-653660451 and that would be it. We will be testing these changes in my server so if I broke anything, we should catch it.

**Comment by TerryGeng** [comment]
Sorry for the delayed response. I will look into this matter in a few hours.

**Comment by felix91gr** [comment]
It's okay @TerryGeng, I just spammed you guys because I was focusing a lot these last two days, and got a lot done as well because I had the time to do it. Take your time, go at your own pace ❤️

**Comment by felix91gr** [comment]
PS: for whenever you have the time, [here](https://www.w3.org/WAI/tutorials/page-structure/regions/) is a good explanation of page regions. Maybe we could use a `header` at least for the banner at the top, now that I think more about it.

**Comment by felix91gr** [comment]
@TerryGeng I have made the banner a `<header>` and the playlist a `<main>`. Visually, they do not change, but now they convey their roles more meaningfully :)

**Comment by felix91gr** [comment]
(Tell me what you think)

**Comment by felix91gr** [comment]
I have added the missing alt text for cover images. I think I'm finally done. So far, I haven't been able to find any more non-contrast errors with WAVE :)

**Comment by TerryGeng** [comment]
Okay. I have looked into the page region and it seems that for our webpage, it contains only a "header" (which is in fact a logo) and the main body. I think what you have done looks good to me, let go with it.

I will check locally to see if anything was broken in the middle and then merge it. Do you think I can squash all commits in this PR into an individual commit?

File: web/templates/index.html [hunk 4]
```diff
@@ -255,50 +285,54 @@ <h5 class="card-title">Filters</h5>
...
-                      <div class="modal-body">
-                        All files listed here, include files on other pages, will be deleted from your
-                        hard-drive.
-                        Is that what you want?
-                      </div>
-                      <div class="modal-footer">
-                        <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
-                        <button id="library-delete-btn" type="button" class="btn btn-danger" data-dismiss="modal">Delete All Listed Files</button>
-                      </div>
+                <!-- QUESTION: should this div have aria-hidden as true?? -->
```
Comment (TerryGeng) [inline web/templates/index.html:313]: I think the answer is "no" since it has `display: none` set unless being triggered, which means it won't be included in the accessibility tree by default.

File: web/templates/index.html [hunk 5]
```diff
@@ -469,18 +511,20 @@ <h5 class="modal-title" id="addTagModalTitle">Edit tags for ?</h5>
...
                     <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
-                    <button id="addTagModalSubmit" type="button" class="btn btn-success" data-dismiss="modal">Edit!</button>
+                    <button id="addTagModalSubmit" type="button" class="btn btn-success"
+                        data-dismiss="modal">Edit!</button>
                 </div>
             </div>
         </div>
     </div>
 
+    <!-- QUESTION: should this div have aria-hidden=true? -->
```
Comment (TerryGeng) [inline web/templates/index.html:521]: See the above comment.

File: web/templates/index.html [hunk 6]
```diff
@@ -140,19 +153,19 @@ <h1>botamusique Web Interface</h1>
...
         <div class="card mb-3">
             <div class="card-header">
-                <h5 class="card-title">Filters</h5>
+                <h3 class="card-title">Filters</h3>
                 <hr>
                 <div class="row">
                     <div class="col">
-                        <label>Type</label>
-                        <div id="filter-type" class="input-group mb-2">
+                        <fieldset id="filter-type" class="input-group mb-2">
```
Comment (TerryGeng) [inline web/templates/index.html:167]: I think you need to remove `input-group` class from this fieldset to avoid stretching these three buttons.

**Comment by felix91gr** [comment]
> Okay. I have looked into the page region and it seems that for our webpage, it contains only a "header" (which is in fact a logo) and the main body. I think what you have done looks good to me, let go with it.
>
> I will check locally to see if anything was broken in the middle and then merge it. 

Sweet!

> Do you think I can squash all commits in this PR into an individual commit?

Sure, go ahead! :)

File: web/templates/index.html (hunk 6 above)
Comment (felix91gr) [reply web/templates/index.html:167]: Stretching? Removing that class on my end (using the web inspector in Firefox) seems to have no effect on the buttons :thinking:

File: web/templates/index.html (hunk 4 above)
Comment (felix91gr) [reply web/templates/index.html:313]: Oh, that's true. Excellent :)

**Comment by TerryGeng** [comment]
Merged. Thanks for your contribution!

**Comment by felix91gr** [comment]
It is a pleasure! 🌈
